charmhelpers/core/hookenv.py

```diff
@@ -510,6 +510,35 @@ def related_units(relid=None):
         subprocess.check_output(units_cmd_line).decode('UTF-8')) or []
 
 
+def expected_related_units(reltype=None):
+    """Get a generator for units we expect to join relation based on
+    goal-state.
+
+    This function will raise KeyError if you request information for a
+    relation type for which juju goal-state does not have information.  It will
+    raise NotImplementedError if used with juju versions without goal-state
+    support.
+
+    :param reltype: Relation type to list data for, default is to list data for
+                    the realtion we are currently executing a hook for.
+    :type reltype: str
+    :returns: iterator
+    :rtype: types.GeneratorType
+    :raises: KeyError, NotImplementedError
+    """
+    if not has_juju_version("2.4.4"):
+        # goal-state existed in 2.4.0, but did not list individual units to
+        # join a relation in 2.4.1 through 2.4.3. (LP: #1794739)
+        raise NotImplementedError("goal_state relation unit count")
+    reltype = reltype or relation_type()
+    _goal_state = goal_state()
+    if reltype == 'cluster':
+        return (key for key in _goal_state['units']
+                if '/' in key and key != local_unit())
```

Comment:
Where does the magic string 'cluster' come from? It isn't documented in the docstring. It is going to fail in any case for charms with a relation named 'cluster' (like the Cassandra charm - it might be quite a common name for peer relations).

I think it would be better if you remove this special case, and have this method only return units in the specified relation. A separate expected_units function could return all related units.

Reply:
The intent was to list peer units, which is what the `units` key in the `goal-state` output contains, not all related units.

I do agree that hiding this under the `cluster` key is odd and that this probably warrants a separate function.

All peer relations of the OpenStack charms are called `cluster` which is where my assumption came from.  Thank you for pointing it out.